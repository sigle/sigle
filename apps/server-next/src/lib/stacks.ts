import { STACKS_MAINNET, STACKS_TESTNET, STACKS_DEVNET } from '@stacks/network';
import { createClient } from '@stacks/blockchain-api-client';
import { env } from '~/env';

export const stacksNetwork =
  env.STACKS_ENV === 'mainnet'
    ? STACKS_MAINNET
    : env.STACKS_ENV === 'testnet'
      ? STACKS_TESTNET
      : STACKS_DEVNET;

export const stacksApiClient = createClient({
  baseUrl: `https://api.${env.STACKS_ENV}.hiro.so`,
});
