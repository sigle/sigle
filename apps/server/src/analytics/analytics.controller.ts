import { Controller, Get, Query, Request, UseGuards } from '@nestjs/common';
import {
  ApiBearerAuth,
  ApiOkResponse,
  ApiOperation,
  ApiTags,
} from '@nestjs/swagger';
import { Throttle } from '@nestjs/throttler';
import { AuthGuard, AuthenticatedRequest } from '../auth.guard';
import { AnalyticsService } from './analytics.service';
import { HistoricalDto } from './dto/historical.dto';
import { HistoricalQueryDto } from './dto/historicalQuery.dto';
import { ReferrerDto } from './dto/referrers.dto';
import { ReferrersQueryDto } from './dto/referrersQuery.dto';

@ApiTags('analytics')
@Controller()
export class AnalyticsController {
  constructor(private readonly analyticsService: AnalyticsService) {}
  @ApiOperation({
    description: 'Return the referrer statistics.',
  })
  @ApiBearerAuth()
  @ApiOkResponse({
    type: ReferrerDto,
    isArray: true,
    schema: {
      example: [
        { domain: 'twitter.com', count: 5 },
        { domain: 'google.com', count: 2 },
      ],
    },
  })
  @Throttle({ default: { limit: 10, ttl: 60000 } })
  @UseGuards(AuthGuard)
  @Get('/api/analytics/referrers')
  getReferrers(
    @Request() req: AuthenticatedRequest,
    @Query() query: ReferrersQueryDto,
  ): Promise<ReferrerDto[]> {
    return this.analyticsService.referrers({
      stacksAddress: req.user.stacksAddress,
      dateFrom: query.dateFrom,
      storyId: query.storyId,
    });
  }

  @ApiOperation({
    description: 'Return the historical statistics.',
  })
  @ApiBearerAuth()
  @ApiOkResponse({
    type: HistoricalDto,
    schema: {
      example: {
        historical: [
          {
            date: '2022-03',
            visits: 0,
            pageviews: 0,
          },
          {
            date: '2022-04',
            visits: 0,
            pageviews: 0,
          },
        ],
        stories: [
          { pathname: 'zsoVIi3V6CE', visits: 0, pageviews: 0 },
          { pathname: '0jE9PPqbxUp', visits: 0, pageviews: 0 },
        ],
      },
    },
  })
  @Throttle({ default: { limit: 10, ttl: 60000 } })
  @UseGuards(AuthGuard)
  @Get('/api/analytics/historical')
  getHistorical(
    @Request() req: AuthenticatedRequest,
    @Query() query: HistoricalQueryDto,
  ): Promise<HistoricalDto> {
    return this.analyticsService.historical({
      stacksAddress: req.user.stacksAddress,
      dateFrom: query.dateFrom,
      dateGrouping: query.dateGrouping,
      storyId: query.storyId,
    });
  }
}
