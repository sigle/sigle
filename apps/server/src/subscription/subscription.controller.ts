import {
  Controller,
  Get,
  HttpCode,
  Post,
  Request,
  UseGuards,
} from '@nestjs/common';
import {
  ApiBearerAuth,
  ApiOkResponse,
  ApiOperation,
  ApiTags,
} from '@nestjs/swagger';
import { Throttle } from '@nestjs/throttler';
import { AuthGuard, AuthenticatedRequest } from '../auth.guard';
import { SubscriptionDto } from './dto/subscription.dto';
import { SubscriptionService } from './subscription.service';

@ApiTags('subscription')
@Controller()
export class SubscriptionController {
  constructor(private readonly subscriptionService: SubscriptionService) {}

  @ApiOperation({
    description:
      'Return the current active subscription of the current logged in user.',
  })
  @ApiBearerAuth()
  @ApiOkResponse({
    description:
      'Returns the current active subscription object. If no active subscription is found, null is returned.',
    type: SubscriptionDto,
  })
  @UseGuards(AuthGuard)
  @Get('/api/subscriptions')
  getUserMe(
    @Request() req: AuthenticatedRequest,
  ): Promise<SubscriptionDto | null> {
    return this.subscriptionService.getUserActiveSubscription({
      stacksAddress: req.user.stacksAddress,
    });
  }

  @ApiOperation({
    description:
      'Create or update a creator plus subscription on the current logged in user. A user can only have one active subscription at a time.',
  })
  @ApiBearerAuth()
  @ApiOkResponse({
    description: 'Returns the newly created subscription object.',
    type: SubscriptionDto,
  })
  @Throttle({ default: { limit: 5, ttl: 60000 } })
  @UseGuards(AuthGuard)
  @Post('/api/subscriptions/syncWithNft')
  @HttpCode(200)
  syncSubscriptionWithNft(
    @Request() req: AuthenticatedRequest,
  ): Promise<SubscriptionDto | null> {
    return this.subscriptionService.syncSubscriptionWithNft({
      stacksAddress: req.user.stacksAddress,
    });
  }
}
