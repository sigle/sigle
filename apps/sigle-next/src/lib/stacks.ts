import { env } from '@/env';
import { AppConfig, UserSession } from '@stacks/connect';
import { STACKS_MAINNET, STACKS_TESTNET, STACKS_DEVNET } from '@stacks/network';
import { createClient } from '@stacks/blockchain-api-client';

export const appConfig = new AppConfig([]);
export const userSession = new UserSession({ appConfig });

export const stacksNetwork =
  env.NEXT_PUBLIC_STACKS_ENV === 'mainnet'
    ? STACKS_MAINNET
    : env.NEXT_PUBLIC_STACKS_ENV === 'testnet'
      ? STACKS_TESTNET
      : STACKS_DEVNET;

export const stacksApiClient = createClient({
  baseUrl: `https://api.${env.NEXT_PUBLIC_STACKS_ENV}.hiro.so`,
});

export const appDetails = {
  name: 'Sigle',
  icon: 'https://app.sigle.io/icon-192x192.png',
};

export const getExplorerTransactionUrl = (txId: string) =>
  `https://explorer.hiro.so/txid/${txId}?chain=${env.NEXT_PUBLIC_STACKS_ENV}`;

export const formatReadableAddress = (address: string) =>
  `${address.substring(0, 6)}...${address.substring(address.length - 4)}`;
