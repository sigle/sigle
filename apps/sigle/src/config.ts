export const appConfig = {
  docsUrl: "https://sigle-docs-next.vercel.app",
};
