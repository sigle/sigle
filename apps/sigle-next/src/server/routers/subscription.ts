import { NonFungibleTokensApi } from '@stacks/blockchain-api-client';
import { prismaClient } from '@/lib/prisma';
import { postHogClient } from '@/lib/posthog';
import { router, authedProcedure } from '../trpc';

const NUMBER_NFTS_BASIC = 1;
const NUMBER_NFTS_PUBLISHER = 3;

export const subscriptionRouter = router({
  /**
   * Get the current active subscription for the logged in user.
   */
  getActive: authedProcedure.query(async ({ ctx }) => {
    const userId = ctx.session.user.id;
    const activeSubscription = await prismaClient.subscription.findFirst({
      select: {
        id: true,
        plan: true,
      },
      where: {
        userId,
        status: 'ACTIVE',
      },
    });
    return activeSubscription;
  }),

  /**
   * Upgrade the current active subscription based on the number of NFTs the user owns.
   */
  upgradeWithNFT: authedProcedure.mutation(async ({ ctx }) => {
    const userId = ctx.session.user.id;
    const user = await prismaClient.user.findUniqueOrThrow({
      select: {
        address: true,
      },
      where: { id: userId },
    });

    const nftApi = new NonFungibleTokensApi();
    const nftHoldings = await nftApi.getNftHoldings({
      principal: user.address,
      assetIdentifiers: [
        'SP2X0TZ59D5SZ8ACQ6YMCHHNR2ZN51Z32E2CJ173.the-explorer-guild::The-Explorer-Guild',
      ],
    });
    const numberNfts = nftHoldings.total;
    if (numberNfts < NUMBER_NFTS_BASIC) {
      return false;
    }

    const newSubscriptionPlan =
      numberNfts >= NUMBER_NFTS_PUBLISHER ? 'PUBLISHER' : 'BASIC';

    let activeSubscription = await prismaClient.subscription.findFirst({
      select: {
        id: true,
        plan: true,
      },
      where: {
        userId: userId,
        status: 'ACTIVE',
      },
    });
    // No need to update the subscription as it's already active and has the correct plan.
    if (activeSubscription && activeSubscription.plan === newSubscriptionPlan) {
      return false;
    }

    // Mark the current active subscription as cancelled.
    if (activeSubscription) {
      await prismaClient.subscription.update({
        select: {
          id: true,
        },
        where: {
          id: activeSubscription.id,
        },
        data: {
          status: 'INACTIVE',
        },
      });
    }

    const newSubscription = await prismaClient.subscription.create({
      select: {
        id: true,
      },
      data: {
        userId: userId,
        status: 'ACTIVE',
        plan: newSubscriptionPlan,
      },
    });

    postHogClient.capture({
      distinctId: userId,
      event: 'subscription created',
      properties: {
        subscriptionId: newSubscription.id,
        numberNfts,
        plan: newSubscriptionPlan,
        previousSubscriptionId: activeSubscription?.id,
      },
    });

    await postHogClient.shutdownAsync();

    return true;
  }),
});
