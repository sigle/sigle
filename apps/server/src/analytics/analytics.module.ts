import { Module } from '@nestjs/common';
import { StacksService } from '../stacks/stacks.service';
import { SubscriptionService } from '../subscription/subscription.service';
import { AnalyticsController } from './analytics.controller';
import { AnalyticsService } from './analytics.service';
import { PlausibleService } from '../plausible/plausible.service';

@Module({
  controllers: [AnalyticsController],
  providers: [
    StacksService,
    PlausibleService,
    SubscriptionService,
    AnalyticsService,
  ],
  imports: [],
})
export class AnalyticsModule {}
