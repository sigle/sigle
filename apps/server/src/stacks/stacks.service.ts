import { Injectable } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { lookupProfile } from '@stacks/auth';
import {
  migrationStories,
  SettingsFile,
  Story,
  SubsetStory,
} from '../external/gaia';
import { fetch } from 'undici';
import { EnvironmentVariables } from '../environment/environment.validation';

@Injectable()
export class StacksService {
  private readonly publicStoriesFileName = 'publicStories.json';
  private readonly settingsFileName = 'settings.json';

  constructor(
    private readonly configService: ConfigService<EnvironmentVariables>,
  ) {}

  /**
   * Return the username associated to a stacks address.
   * If the user has 1 free subdomain and 1 .btc, the .btc will be returned.
   */
  async getUsernameByAddress(address: string): Promise<string | null> {
    const namesResponse = await fetch(
      `https://api.hiro.so/v1/addresses/stacks/${address}`,
    );
    const namesJson = (await namesResponse.json()) as { names: string[] };
    if (namesJson.names && namesJson.names.length > 0) {
      return (
        namesJson.names.find(
          (name: string) => name.endsWith('.btc') === true,
        ) || namesJson.names[0]
      );
    }
    return null;
  }

  async getBucketUrl({
    username,
  }: {
    username: string;
  }): Promise<{ profile: Record<string, any>; bucketUrl: string }> {
    let userProfile: Record<string, any> | undefined;
    try {
      userProfile = await lookupProfile({ username });
    } catch (error) {
      // This will happen if there is no blockstack user with this name
      if (error?.message === 'Name not found') {
        userProfile = undefined;
      } else {
        throw error;
      }
    }

    const bucketUrl: string | undefined =
      userProfile &&
      userProfile.apps &&
      userProfile.apps[this.configService.get('APP_URL')];

    return { profile: userProfile, bucketUrl };
  }

  async getPublicStories({
    bucketUrl,
  }: {
    bucketUrl: string;
  }): Promise<SubsetStory[]> {
    const resPublicStories = await fetch(
      `${bucketUrl}${this.publicStoriesFileName}`,
    );
    // This would happen if the user has not published any stories
    if (resPublicStories.status !== 200) {
      return [];
    }
    const publicStoriesFile = migrationStories(
      (await resPublicStories.json()) as any,
    );
    return publicStoriesFile.stories;
  }

  async getPublicStory({
    bucketUrl,
    storyId,
  }: {
    bucketUrl: string;
    storyId: string;
  }): Promise<Story | null> {
    const resPublicStory = await fetch(`${bucketUrl}${storyId}.json`);
    if (resPublicStory.status !== 200) {
      return null;
    }
    return (await resPublicStory.json()) as Story;
  }

  async getPublicSettings({
    bucketUrl,
  }: {
    bucketUrl: string;
  }): Promise<SettingsFile> {
    const resPublicSettings = await fetch(
      `${bucketUrl}${this.settingsFileName}`,
    );
    // This would happen if the user does not have any settings
    if (resPublicSettings.status !== 200) {
      return {};
    }
    return await resPublicSettings.json();
  }
}
