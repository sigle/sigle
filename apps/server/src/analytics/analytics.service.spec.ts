import { CACHE_MANAGER } from '@nestjs/cache-manager';
import { ConfigService } from '@nestjs/config';
import { Test, TestingModule } from '@nestjs/testing';
import { PlausibleService } from '../plausible/plausible.service';
import { StacksService } from '../stacks/stacks.service';
import { SubscriptionService } from '../subscription/subscription.service';
import { AnalyticsService } from './analytics.service';

describe('AnalyticsService', () => {
  let service: AnalyticsService;
  let plausibleService: PlausibleService;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [
        AnalyticsService,
        {
          provide: CACHE_MANAGER,
          useValue: {
            get: () => null,
            set: () => () => null,
          },
        },
        {
          provide: ConfigService,
          useValue: {},
        },
        {
          provide: SubscriptionService,
          useValue: {
            getUserActiveSubscription: () => () => ({
              id: 'test',
            }),
          },
        },
        {
          provide: StacksService,
          useValue: {
            getUsernameByAddress: () => 'sigle.btc',
            getBucketUrl: () => ({
              profile: {},
              bucketUrl: 'https://www.sigle.io',
            }),
            getPublicStories: () => [],
          },
        },
        {
          provide: PlausibleService,
          useValue: {
            referrers: jest.fn(),
            timeseries: jest.fn(),
            pages: jest.fn(),
          },
        },
      ],
    }).compile();

    service = module.get<AnalyticsService>(AnalyticsService);
    plausibleService = await module.resolve(PlausibleService);
  });

  describe('referrers', () => {
    it('should return referrers', async () => {
      jest
        .useFakeTimers()
        .setSystemTime(new Date('2022-04-04 12:00:00 UTC').getTime());
      const stacksAddress = 'SP2BKHGRV8H2YDJK16FP5VASYFDGYN4BPTNFWDKYJ';
      (plausibleService.referrers as jest.Mock).mockResolvedValueOnce([
        {
          pageviews: '1',
          source: 'https://stacks.co',
        },
        {
          pageviews: '10',
          source: 'Direct / None',
        },
      ]);

      const data = await service.referrers({
        stacksAddress,
        dateFrom: '2022-03-15',
      });
      expect(plausibleService.referrers).toBeCalledWith({
        dateFrom: '2022-05-01',
        dateTo: '2022-04-04',
        paths: expect.any(Array),
      });
      expect(data).toMatchSnapshot();
    });

    it('should return referrers for one story', async () => {
      jest
        .useFakeTimers()
        .setSystemTime(new Date('2022-04-04 12:00:00 UTC').getTime());
      const stacksAddress = 'SP2BKHGRV8H2YDJK16FP5VASYFDGYN4BPTNFWDKYJ';
      (plausibleService.referrers as jest.Mock).mockResolvedValueOnce([
        {
          pageviews: '5',
          source: 'https://stacks.co',
        },
      ]);

      const data = await service.referrers({
        stacksAddress,
        dateFrom: '2022-03-15',
        storyId: 'test',
      });
      expect(plausibleService.referrers).toBeCalledWith({
        dateFrom: '2022-05-01',
        dateTo: '2022-04-04',
        paths: ['/sigle.btc/test'],
      });
      expect(data).toMatchSnapshot();
    });
  });

  describe('historical', () => {
    it('should return a formatted time series for days', async () => {
      jest
        .useFakeTimers()
        .setSystemTime(new Date('2022-04-04 12:00:00 UTC').getTime());
      const stacksAddress = 'SP2BKHGRV8H2YDJK16FP5VASYFDGYN4BPTNFWDKYJ';
      (plausibleService.timeseries as jest.Mock).mockResolvedValueOnce([
        {
          visitors: '5',
          pageviews: '1',
          date: '2022-03-20',
        },
        {
          visitors: '7',
          pageviews: '3',
          date: '2022-03-20',
        },
      ]);
      (plausibleService.pages as jest.Mock).mockResolvedValueOnce([
        {
          visitors: '5',
          pageviews: '1',
          page: '/test.btc/path-test',
        },
        {
          visitors: '7',
          pageviews: '3',
          page: '/test.btc/path-test-2',
        },
      ]);

      const data = await service.historical({
        stacksAddress,
        dateFrom: '2022-03-15',
        dateGrouping: 'day',
      });
      expect(plausibleService.timeseries).toBeCalledWith({
        dateFrom: '2022-05-01',
        dateGrouping: 'day',
        dateTo: '2022-04-04',
        paths: expect.any(Array),
      });
      expect(plausibleService.pages).toBeCalledWith({
        dateFrom: '2022-05-01',
        dateTo: '2022-04-04',
        paths: expect.any(Array),
      });
      expect(data).toMatchSnapshot();
    });

    it('should return a formatted time series for months', async () => {
      jest
        .useFakeTimers()
        .setSystemTime(new Date('2022-04-04 12:00:00 UTC').getTime());
      const stacksAddress = 'SP2BKHGRV8H2YDJK16FP5VASYFDGYN4BPTNFWDKYJ';
      (plausibleService.timeseries as jest.Mock).mockResolvedValueOnce([
        {
          visitors: '5',
          pageviews: '1',
          date: '2022-03-01',
        },
      ]);
      (plausibleService.pages as jest.Mock).mockResolvedValueOnce([
        {
          visitors: '5',
          pageviews: '1',
          page: '/test.btc/path-test',
        },
        {
          visitors: '7',
          pageviews: '3',
          page: '/test.btc/path-test-2',
        },
      ]);

      const data = await service.historical({
        stacksAddress,
        dateFrom: '2022-01-02',
        dateGrouping: 'month',
      });
      expect(plausibleService.timeseries).toBeCalledWith({
        dateFrom: '2022-05-01',
        dateGrouping: 'day',
        dateTo: '2022-04-04',
        paths: expect.any(Array),
      });
      expect(plausibleService.pages).toBeCalledWith({
        dateFrom: '2022-05-01',
        dateTo: '2022-04-04',
        paths: expect.any(Array),
      });
      expect(data).toMatchSnapshot();
    });

    it('should return a formatted time series for one story', async () => {
      jest
        .useFakeTimers()
        .setSystemTime(new Date('2022-04-04 12:00:00 UTC').getTime());
      const stacksAddress = 'SP2BKHGRV8H2YDJK16FP5VASYFDGYN4BPTNFWDKYJ';
      (plausibleService.timeseries as jest.Mock).mockResolvedValueOnce([
        {
          visitors: '5',
          pageviews: '1',
          date: '2022-03-01',
        },
      ]);
      (plausibleService.pages as jest.Mock).mockResolvedValueOnce([
        {
          visitors: '5',
          pageviews: '1',
          page: '/test.btc/path-test',
        },
      ]);

      const data = await service.historical({
        stacksAddress,
        dateFrom: '2022-03-15',
        dateGrouping: 'day',
        storyId: 'test',
      });
      expect(plausibleService.timeseries).toBeCalledWith({
        dateFrom: '2022-05-01',
        dateGrouping: 'day',
        dateTo: '2022-04-04',
        paths: expect.any(Array),
      });
      expect(plausibleService.pages).toBeCalledWith({
        dateFrom: '2022-05-01',
        dateTo: '2022-04-04',
        paths: expect.any(Array),
      });
      expect(data).toMatchSnapshot();
    });
  });
});
