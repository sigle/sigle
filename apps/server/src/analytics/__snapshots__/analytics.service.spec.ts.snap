// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnalyticsService historical should return a formatted time series for days 1`] = `
{
  "historical": [
    {
      "date": "2022-03-20",
      "pageviews": "1",
      "visits": "5",
    },
    {
      "date": "2022-03-20",
      "pageviews": "3",
      "visits": "7",
    },
  ],
  "stories": [
    {
      "pageviews": "1",
      "pathname": "/test.btc/path-test",
      "visits": "5",
    },
    {
      "pageviews": "3",
      "pathname": "/test.btc/path-test-2",
      "visits": "7",
    },
  ],
}
`;

exports[`AnalyticsService historical should return a formatted time series for months 1`] = `
{
  "historical": [
    {
      "date": "2022-03-01",
      "pageviews": "1",
      "visits": "5",
    },
  ],
  "stories": [
    {
      "pageviews": "1",
      "pathname": "/test.btc/path-test",
      "visits": "5",
    },
    {
      "pageviews": "3",
      "pathname": "/test.btc/path-test-2",
      "visits": "7",
    },
  ],
}
`;

exports[`AnalyticsService historical should return a formatted time series for one story 1`] = `
{
  "historical": [
    {
      "date": "2022-03-01",
      "pageviews": "1",
      "visits": "5",
    },
  ],
  "stories": [
    {
      "pageviews": "1",
      "pathname": "/test.btc/path-test",
      "visits": "5",
    },
  ],
}
`;

exports[`AnalyticsService referrers should return referrers 1`] = `
[
  {
    "count": "10",
    "domain": "Direct / None",
  },
  {
    "count": "1",
    "domain": "https://stacks.co",
  },
]
`;

exports[`AnalyticsService referrers should return referrers for one story 1`] = `
[
  {
    "count": "5",
    "domain": "https://stacks.co",
  },
]
`;
